study1/week2/range-sum-of-bst.java

```diff
@@ -0,0 +1,41 @@
+//https://leetcode.com/problems/range-sum-of-bst/description/
+
+/**
+ * Definition for a binary tree node.
+ * public class TreeNode {
+ *     int val;
+ *     TreeNode left;
+ *     TreeNode right;
+ *     TreeNode() {}
+ *     TreeNode(int val) { this.val = val; }
+ *     TreeNode(int val, TreeNode left, TreeNode right) {
+ *         this.val = val;
+ *         this.left = left;
+ *         this.right = right;
+ *     }
+ * }
+ */
+class Solution {
+
+    public int rangeSumBST(TreeNode root, int low, int high) {
+        //time complexity is O(n), because visits n nodes to check
+        //space complexity is O(1), bacause there is no new structure
```

Comment:
When you say `no new structure`, what would that exactly mean?

Reply:
Also I thought this was correct, but as you're adding to the stack each call. in case this is an unbalanced BST, it will be O(n) because at maximum the call stack will contain n calls at the same time.

Reply:
hmmmm actually I thought the space complexity had a effect on the structure of creating a new things (ex. `new int[n]`). So, in my code, I thought that there was no new structure itself, so I thought it was O(1). But did I get it wrong? 
so,,, I don't actually understand your explanation of O(n)😔

Reply:
So I guess you understand that there will be `n` calls in the stack in the worst case.
in those stack, it will have space for the `sum` variable on each call in the stack. Thus there will be `n` sum variables in the worst case. meaning that the space complexity is `O(n)` because you've allocated `n` sum variables on stack.

It doesn't matter for space complexity whether you've allocated on heap by calling `new XX()` or on stack. The space complexity should take into account the maximum number of allocation happening irrespective of the type of memory.